File: vi/codes/ruby/chapter_backtracking/subset_sum_i_naive.rb
=begin
File: subset_sum_i_naive.rb
Created Time: 2024-05-22
Author: Xuan Khoa Tu Nguyen (ngxktuzkai2000@gmail.com)
=end

### Quay lui: tổng tập con I ###
def backtrack(state, target, total, choices, res)
  # Khi tổng của tập con bằng target, ghi lại lời giải
  if total == target
    res << state.dup
    return
  end

  # Duyệt qua tất cả các lựa chọn
  for i in 0...choices.length
    # Cắt tỉa: nếu tổng của tập con vượt quá target, bỏ qua lựa chọn này
    next if total + choices[i] > target
    # Thử: thực hiện lựa chọn, cập nhật tổng phần tử total
    state << choices[i]
    # Tiến hành vòng lựa chọn tiếp theo
    backtrack(state, target, total + choices[i], choices, res)
    # Quay lui: hoàn tác lựa chọn, khôi phục trạng thái trước đó
    state.pop
  end
end

### Giải bài toán tổng tập con I (có tập con trùng lặp) ###
def subset_sum_i_naive(nums, target)
  state = [] # Trạng thái (tập con)
  total = 0 # Tổng của tập con
  res = [] # Danh sách kết quả (danh sách các tập con)
  backtrack(state, target, total, nums, res)
  res
end

### Mã chương trình chính ###
if __FILE__ == $0
  nums = [3, 4, 5]
  target = 9
  res = subset_sum_i_naive(nums, target)

  puts "Input array nums = #{nums}, target = #{target}"
  puts "All subsets with sum equal to #{target} res = #{res}"
  puts "Please note that this method outputs results containing duplicate sets"
end

File: vi/codes/rust/chapter_backtracking/subset_sum_i_naive.rs
/*
 * File: subset_sum_i_naive.rs
 * Created Time: 2023-07-09
 * Author: codingonion (coderonion@gmail.com)
 */

/* Giải thuật quay lui: Tổng tập con I */
fn backtrack(
    state: &mut Vec<i32>,
    target: i32,
    total: i32,
    choices: &[i32],
    res: &mut Vec<Vec<i32>>,
) {
    // Khi tổng tập con bằng target, ghi lại nghiệm
    if total == target {
        res.push(state.clone());
        return;
    }
    // Duyệt qua tất cả các lựa chọn
    for i in 0..choices.len() {
        // Cắt tỉa: nếu tổng tập con vượt quá target, bỏ qua lựa chọn này
        if total + choices[i] > target {
            continue;
        }
        // Thử: thực hiện lựa chọn, cập nhật tổng phần tử total
        state.push(choices[i]);
        // Tiến hành vòng lựa chọn tiếp theo
        backtrack(state, target, total + choices[i], choices, res);
        // Quay lui: hoàn tác lựa chọn, khôi phục về trạng thái trước đó
        state.pop();
    }
}

/* Giải bài toán tổng tập con I (bao gồm các tập con trùng lặp) */
fn subset_sum_i_naive(nums: &[i32], target: i32) -> Vec<Vec<i32>> {
    let mut state = Vec::new(); // Trạng thái (tập con)
    let total = 0; // Tổng tập con
    let mut res = Vec::new(); // Danh sách kết quả (danh sách các tập con)
    backtrack(&mut state, target, total, nums, &mut res);
    res
}

/* Mã điều khiển chính */
pub fn main() {
    let nums = [3, 4, 5];
    let target = 9;

    let res = subset_sum_i_naive(&nums, target);

    println!("Input array nums = {:?}, target = {}", &nums, target);
    println!("All subsets with sum equal to {} res = {:?}", target, &res);
    println!("Please note that this method outputs results containing duplicate sets");
}
